titiler/cmr/timeseries.py

```diff
@@ -0,0 +1,588 @@
+        ),
+    ] = None
+    step_idx: Annotated[
+        Optional[int],
+        Query(description="Optional (zero-indexed) index of the desired time step"),
+    ] = None
+    exact: Annotated[
+        Optional[bool],
+        Query(
+            description="If true, queries will be made for a point-in-time at each step. If false, queries will be made for the entire interval between steps"
+        ),
+    ] = None
+    datetimes: Annotated[
+        Optional[str],
+        Query(
+            description="Optional list of comma-separated specific time points or time intervals to summarize over"
+        ),
+    ] = None
+
+
+def parse_duration(duration: str) -> relativedelta:
+    """Parse ISO 8601 duration string to relativedelta."""
+    match = re.match(
+        r"P(?:(\d+)Y)?(?:(\d+)M)?(?:(\d+)W)?(?:(\d+)D)?(?:T(?:(\d+)H)?(?:(\d+)M)?(?:(\d+(?:\.\d+)?)S)?)?",
+        duration,
+    )
+    if not match or not any(m for m in match.groups()):
+        raise ValueError(f"{duration} is an invalid duration format")
+
+    years, months, weeks, days, hours, minutes, seconds = [
+        float(g) if g else 0 for g in match.groups()
+    ]
+    return relativedelta(
+        years=int(years),
+        months=int(months),
+        weeks=int(weeks),
+        days=int(days),
+        hours=int(hours),
+        minutes=int(minutes),
+        seconds=int(seconds),
+        microseconds=int((seconds % 1) * 1e6),
+    )
+
+
+def generate_datetime_ranges(
+    start_datetime: datetime, end_datetime: datetime, step: str, exact: bool = False
+) -> List[Union[Tuple[datetime], Tuple[datetime, datetime]]]:
+    """Generate datetime ranges"""
+    start = start_datetime
+    end = end_datetime
+    step_delta = parse_duration(step)
+
+    ranges: List[Union[Tuple[datetime], Tuple[datetime, datetime]]] = []
+    current = start
+
+    step_timedelta = (current + step_delta) - current
+    is_small_timestep = step_timedelta <= timedelta(seconds=1)
+
+    while current < end:
+        if exact:
+            # For exact case, return a tuple with just one exact datetime
+            next_step = current + step_delta
+            ranges.append((current,))
+        else:
+            next_step = min(current + step_delta, end)
+            if next_step == end:
+                ranges.append((current, next_step))
+                break
+
+            if is_small_timestep:
+                # Subtract 1 millisecond for small timesteps
+                ranges.append((current, next_step - timedelta(microseconds=1)))
+            else:
+                # Subtract 1 second for larger timesteps
+                ranges.append((current, next_step - timedelta(seconds=1)))
+
+        current = next_step
+
+        if current == end:
+            ranges.append((end,))
+
+    if not ranges:
+        return [(start, end)]
+
+    return ranges
```

Comment:
All of this would eventually move to `titiler.extensions.timeseries` but I wanted to do the development all in one place to facilitate easier iteration during early stages of development.